include/seastar/core/internal/cpu_profiler.hh

```diff
@@ -0,0 +1,154 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2023 ScyllaDB
+ */
+
+#pragma once
+
+#include <seastar/core/circular_buffer_fixed_capacity.hh>
+#include <seastar/core/internal/timers.hh>
+#include <seastar/util/backtrace.hh>
+
+#include <boost/container/static_vector.hpp>
+#include <chrono>
+#include <csignal>
+#include <cstdint>
+#include <ctime>
+#include <mutex>
+#include <atomic>
+#include <optional>
+#include <random>
+
+namespace seastar {
+
+class reactor;
+
+struct cpu_profiler_trace {
+    using kernel_trace_vec = boost::container::static_vector<uintptr_t, 64>;
+    simple_backtrace user_backtrace;
+    kernel_trace_vec kernel_backtrace;
+};
+
+constexpr size_t max_number_of_traces = 128;
+
+namespace internal {
+
+/// A lightweight mutex designed to work with interrupts
+/// utilizing only compiler barriers.
+class signal_mutex {
+public:
+    class guard {
+    private:
+        signal_mutex* _mutex;
+        guard(signal_mutex* m) : _mutex(m) {}
+        friend class signal_mutex;
+    public:
+        guard(guard&& o) : _mutex(o._mutex) { o._mutex = nullptr; }
+        ~guard();
+    };
+
+    // Returns a `guard` if the lock was acquired.
+    // Otherwise returns a nullopt.
+    std::optional<guard> try_lock();
+
+private:
+    friend class guard;
+    std::atomic_bool _mutex;
+};
+
+struct cpu_profiler_config {
+    bool enabled;
+    std::chrono::nanoseconds period;
+};
+
+class cpu_profiler {
+private:
+    circular_buffer_fixed_capacity<cpu_profiler_trace, max_number_of_traces> _traces;
+    // The operations in `_traces` are not reentrant. Therefore mutex is used to ensure
+    // that an interrupt cannot access `_traces` if the interrupted thread was already
+    // accessing it.
+    signal_mutex _traces_mutex;
+    cpu_profiler_config _cfg;
+    std::chrono::nanoseconds _last_set_timeout;
+
+
+    bool is_enabled() const;
+    std::chrono::nanoseconds period() const;
+    std::chrono::nanoseconds get_next_timeout();
+
+protected:
+    friend reactor;
+
+public:
+    static int signal_number() { return SIGRTMIN + 2; }
+
+    cpu_profiler(cpu_profiler_config cfg) : _cfg(cfg) {}
+
+    void update_config(cpu_profiler_config cfg);
+    void stop();
```

Comment:
Good point, I'll add comments to each of them.

As for why `start/stop` instead of just `update_config`. It's mainly since `update_config` is user-driven and defined. There are times when the reactor needs to control the profiler as well. Like when going to sleep. If the only way of doing that was `update_config` it'd be a rather tedious process of saving the current config, updating the enable/disable value to what the reactor wanted, then restoring the original config afterwards.

That being said now that I've thought about it I need to change `update_config` not to start the profiler again if something non-user-configured stopped it to begin with. Will make that change as well.

Reply:
Makes sense: I also wrote that before I saw down below how start() / stop() are used in the same way as begin_sleep / end_sleep for the stall notifier.